File: exercise-03-build-your-own-tooling/mcp-exercises/nodejs-starter/src/features/users/bad-code.ts
// ============================================================
// EXERCISE 4 — REFACTOR TARGET
//
// This file intentionally violates every Node.js coding standard.
// Ask the AI to refactor it — first WITHOUT the MCP server,
// then WITH it. Compare how thoroughly each version fixes the issues.
//
// Violations in this file:
//   - PascalCase function names (should be camelCase with verb prefix)
//   - Uses axios (forbidden — use native fetch)
//   - Uses .then() chains (forbidden — use async/await)
//   - Uses `any` type everywhere (forbidden — use proper types)
//   - Raw throw for control flow (forbidden — use Result pattern)
//   - Magic numbers (5, 1800000) — should be named constants
//   - console.log in production code (forbidden — use structured logger)
//   - Hard-coded singleton imports instead of dependency injection
// ============================================================

import axios from "axios";

const db = {
  users: [
    { id: "1", name: "Alice Johnson", email: "alice@example.com", loginAttempts: 0 },
    { id: "2", name: "Bob Smith", email: "bob@example.com", loginAttempts: 6 },
  ],
  findById: (id: string) => db.users.find((u) => u.id === id),
};

// ❌ PascalCase function — should be camelCase with a verb prefix
export async function GetUser(id: any): Promise<any> {
  // ❌ .then() chain — should be async/await
  // ❌ `any` type — should define a User interface
  return Promise.resolve(db.users).then((users: any) => {
    return users.find((u: any) => u.id === id);
  });
}

// ❌ No verb in name — should be getAllUsers or listUsers
// ❌ axios — should use native fetch
export async function UsersList(): Promise<any> {
  // ❌ console.log — should use structured logger
  console.log("fetching users from remote");
  // ❌ axios — forbidden
  // ❌ .then() chain
  return axios.get("https://jsonplaceholder.typicode.com/users").then((res: any) => res.data);
}

// ❌ PascalCase — should be loginUser
// ❌ `any` return type — should define a result shape
export async function LoginUser(email: string, password: string): Promise<any> {
  const user = db.users.find((u) => u.email === email);

  // ❌ raw throw for expected business logic — should return Result type
  if (!user) {
    throw new Error("User not found");
  }

  // ❌ magic number 5 — should be const MAX_LOGIN_ATTEMPTS = 5
  if (user.loginAttempts >= 5) {
    throw new Error("Account locked");
  }

  // ❌ console.log — should use structured logger
  // ❌ string concatenation in log — should use structured context object
  console.log("User logged in: " + user.id);

  // ❌ magic number 1800000 — should be const SESSION_TIMEOUT_MS = 30 * 60 * 1000
  return { token: "mock-token-abc123", expiresIn: 1800000 };
}

File: exercise-03-build-your-own-tooling/mcp-exercises/python-starter/app/users/bad_code.py
# ============================================================
# EXERCISE 4 — REFACTOR TARGET
#
# This file intentionally violates every Python coding standard.
# Ask the AI to refactor it — first WITHOUT the MCP server,
# then WITH it. Compare how thoroughly each version fixes the issues.
#
# Violations in this file:
#   - camelCase function names (should be snake_case with verb prefix)
#   - Uses `requests` (forbidden — use httpx with async)
#   - Synchronous I/O (forbidden — use async/await)
#   - Missing type hints on all functions
#   - Bare `except:` (forbidden — catch specific exception types)
#   - Silenced errors with `pass` (forbidden — never swallow errors)
#   - Magic numbers (5, 1800) — should be named UPPER_SNAKE_CASE constants
#   - print() statements (forbidden — use structlog)
#   - f-string in log call (forbidden — use keyword args)
#   - Hard-coded singleton imports instead of dependency injection
# ============================================================

import requests

from app.database import db  # type: ignore  (not a real module — for demo only)


# ❌ camelCase — should be snake_case
# ❌ No type hints
def GetUser(id):
    # ❌ requests (sync, forbidden) — should use httpx async
    # ❌ bare except — should catch specific exception types
    try:
        response = requests.get(f"/api/users/{id}")
        return response.json()
    except:   # ❌ bare except
        pass  # ❌ silenced error — should raise or log and re-raise


# ❌ camelCase, no verb prefix — should be list_users or get_all_users
# ❌ No type hints
def UsersList():
    # ❌ print — should use structlog
    print("fetching users from remote")
    # ❌ requests (sync, forbidden)
    return requests.get("https://jsonplaceholder.typicode.com/users").json()


# ❌ camelCase — should be login_user
# ❌ No type hints on parameters or return
def LoginUser(email, password):
    users = db.get_all_users()
    user = next((u for u in users if u["email"] == email), None)

    # ❌ raises generic Exception — should raise custom UserNotFoundError
    if not user:
        raise Exception("User not found")

    # ❌ magic number 5 — should be MAX_LOGIN_ATTEMPTS = 5
    if user["loginAttempts"] >= 5:
        raise Exception("Account locked")

    # ❌ print — should use structlog
    # ❌ f-string in log — should use keyword arguments: logger.info("user_logged_in", user_id=user["id"])
    print(f"User logged in: {user['id']}")

    # ❌ magic number 1800 — should be SESSION_TIMEOUT_SECONDS = 30 * 60
    return {"token": "mock-token-abc123", "expires_in": 1800}
